### PR TITLE
Prevent dropping of critical system views, functions, and procedures

### PR DESCRIPTION
**Pull request by roshan0708**
### Description

This commit introduces restrictions to prevent the dropping of essential system views, functions, and procedures within the dbo schema for both TSQL and PSQL dialects:
- sp_addlinkedserver
- syspolicy_configuration
- syspolicy_system_health_state
- fn_syspolicy_is_automation_enabled

Generalized the restriction mechanism for handling all kinds of objects.

Note: Only superusers can drop these restricted objects and since this couldn't be added in jdbc test, adding the output here for more info:
```
-- psql
babelfish_db=# DROP VIEW msdb_dbo.syspolicy_system_health_state;
DROP VIEW
```
### Issues Resolved
JIRA-5145, JIRA-5146

### Test Scenarios Covered ###
* **Use case based -** y

* **Boundary conditions -**

* **Arbitrary inputs -**

* **Negative test cases -**

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/pltsql.h [hunk 1]
```diff
@@ -2348,4 +2348,29 @@ extern bool validate_special_function(char *proc_nsname, char *proc_name, int na
...
+	{"sp_addlinkedserver",		OBJECT_PROCEDURE},
+	{"sp_addlinkedsrvlogin",	OBJECT_PROCEDURE},
+	{"sp_droplinkedsrvlogin",	OBJECT_PROCEDURE},
+	{"sp_dropserver",			OBJECT_PROCEDURE},
+	{"sp_enum_oledb_providers",	OBJECT_PROCEDURE},
+	{"sp_testlinkedserver",		OBJECT_PROCEDURE}
+};
+
+static const RestrictedObject msdb_dbo_objects[] = {
+	{"syspolicy_configuration",				OBJECT_VIEW},
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/src/pltsql.h]: We are also creating `dbo.sysdatabases` view as per of database create, I think we should restrict that as well.

Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/pltsql.h]: This is already restricted: 
```
1> drop view sys.sysdatabases
2> go
Msg 33557097, Level 16, State 1, Server BABELFISH, Line 1
must be owner of view sysdatabases
1> 
```

File: contrib/babelfishpg_tsql/src/pltsql.h [hunk 2]
```diff
@@ -2348,4 +2348,29 @@ extern bool validate_special_function(char *proc_nsname, char *proc_name, int na
...
+	{"xp_instance_regread",		OBJECT_PROCEDURE},
+	{"sp_addlinkedserver",		OBJECT_PROCEDURE},
+	{"sp_addlinkedsrvlogin",	OBJECT_PROCEDURE},
+	{"sp_droplinkedsrvlogin",	OBJECT_PROCEDURE},
+	{"sp_dropserver",			OBJECT_PROCEDURE},
+	{"sp_enum_oledb_providers",	OBJECT_PROCEDURE},
+	{"sp_testlinkedserver",		OBJECT_PROCEDURE}
+};
+
+static const RestrictedObject msdb_dbo_objects[] = {
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/src/pltsql.h]: I would suggest instead of having list for each schema, have something which can be extended to other schema if needed.

File: contrib/babelfishpg_tsql/src/pltsql.h [hunk 3]
```diff
@@ -2348,4 +2348,29 @@ extern bool validate_special_function(char *proc_nsname, char *proc_name, int na
...
+	{"syspolicy_configuration",				OBJECT_VIEW},
+	{"syspolicy_system_health_state",		OBJECT_VIEW},
+	{"fn_syspolicy_is_automation_enabled",	OBJECT_FUNCTION}
+};
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/src/pltsql.h]: I think we are already maintaining these names inside [is_ms_shipped](https://github.com/babelfish-for-postgresql/babelfish_extensions/blame/BABEL_5_X_DEV/contrib/babelfishpg_tsql/runtime/functions.c#L3838) function, which can be declare globally and used here as well.

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 4]
```diff
@@ -6978,6 +6978,10 @@ bbf_ExecDropStmt(DropStmt *stmt)
 
 			if (!relation)
 				continue;
+			
+			/* Restrict dropping of extended stored procedures for non-superuser roles */
+			if (stmt->removeType == OBJECT_VIEW && !superuser())
+				check_restricted_object(address.objectId, OBJECT_VIEW);
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/src/pl_handler.c:7267]: Not related to this but I think we should have OidIsValid check for objectId before this. Please refer to procedure scenario below.

File: contrib/babelfishpg_tsql/src/pl_funcs-2.c [hunk 5]
```diff
@@ -351,58 +352,111 @@ coalesce_typmod_hook_impl(const CoalesceExpr *cexpr)
...
-				is_restricted = true;
-				break;
-			}
-		}
+		case OBJECT_PROCEDURE:
+			tuple = SearchSysCache1(PROCOID, ObjectIdGetDatum(object_id));
+			break;
+		case OBJECT_VIEW:
+			tuple = SearchSysCache1(RELOID, ObjectIdGetDatum(object_id));
+			break;
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/src/pl_funcs-2.c]: Please free this pointer at the end.

File: contrib/babelfishpg_tsql/src/pl_funcs-2.c [hunk 6]
```diff
@@ -351,58 +352,111 @@ coalesce_typmod_hook_impl(const CoalesceExpr *cexpr)
...
+		case OBJECT_FUNCTION:
+			tuple = SearchSysCache1(PROCOID, ObjectIdGetDatum(object_id));
+			break;
+		default:
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("unsupported object type")));
+			return;
 	}
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/src/pl_funcs-2.c:390]: You don't need to release the tuple here since you're going to raise an error and error handler will handle that.

File: contrib/babelfishpg_tsql/src/pltsql.h (hunk 1 above)
Comment (sumitj824) [reply contrib/babelfishpg_tsql/src/pltsql.h]: Please check it from PSQL endpoint as well. We need to handle that also.
```
babelfish_db=> drop view db3_dbo.sysdatabases;
DROP VIEW
```

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 4 above)
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/pl_handler.c:7267]: Will add a check in the follow up commit.

File: contrib/babelfishpg_tsql/src/pltsql.h (hunk 2 above)
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/pltsql.h]: I considered a hash table but stuck with arrays since we’ve got a small number of restricted objects. It keeps things simple and performs well. If things change, we can always tweak it later.

File: contrib/babelfishpg_tsql/src/pltsql.h (hunk 1 above)
Comment (sumitj824) [reply contrib/babelfishpg_tsql/src/pltsql.h]: Please address this.

File: contrib/babelfishpg_tsql/src/pl_funcs-2.c [hunk 7]
```diff
@@ -351,58 +352,78 @@ coalesce_typmod_hook_impl(const CoalesceExpr *cexpr)
...
+	else
+	{
+		Form_pg_class classform = (Form_pg_class) GETSTRUCT(tuple);
+		objname = NameStr(classform->relname);
+		schema_oid = classform->relnamespace;
+	}
+
+	/* Get schema name */
+	if (OidIsValid(schema_oid))
+		schema_name = get_namespace_name(schema_oid);
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/src/pl_funcs-2.c:412]: Please free this pointer.

File: contrib/babelfishpg_tsql/src/pl_funcs-2.c [hunk 8]
```diff
@@ -351,58 +352,84 @@ coalesce_typmod_hook_impl(const CoalesceExpr *cexpr)
...
+		if (schema_name)
+			pfree(schema_name);
 		ereport(ERROR,
 				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
-				 errmsg("must be owner of procedure %s", procname)));
+				 errmsg("must be owner of %s %s",
+				 		 object_type == OBJECT_PROCEDURE ? "procedure" :
+						 object_type == OBJECT_VIEW ? "view" :
+						 "function",
+						 objname)));
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/src/pl_funcs-2.c:430]: A suggestion: This error message is confusing since the owner of these system objects is 'sysadmin' and we won't allow sysadmin(or its member) also to drop these objects either.  "must be owner" is a generic message that comes from PG, lets make it Babelfish specific.
Better error message would be: `Babelfish-created system view/procedure/function cannot be dropped outside of a Babelfish session` similar to what we throw for BBF roles.

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 9]
```diff
@@ -7081,6 +7081,13 @@ bbf_ExecDropStmt(DropStmt *stmt)
 			if (!relation)
 				continue;
 
+			if (!OidIsValid(address.objectId))
+				continue;
+			
+			/* Restrict dropping of system views for non-superuser roles */
+			if (stmt->removeType == OBJECT_VIEW && !superuser())
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: Can we please move `!superuser()` condition inside `check_restricted_object`?

File: test/JDBC/expected/restricted_objects-vu-verify.out [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+EXEC dbo.babel_5146_prepare_p3;
+GO
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: must be owner of view syspolicy_system_health_state)~~
+
+
+-- psql user=babel_5146_user_l1 password=abc
+-- should throw error
+DROP FUNCTION msdb_dbo.fn_syspolicy_is_automation_enabled;
```
Comment (HarshLunagariya) [inline test/JDBC/expected/restricted_objects-vu-verify.out:48]: 1. Please add all 11 object test cases explicitly.
2. Add a testcase when current user is dbo/or member of db_owner/db_ddladmin as well.
3. Add a testcase for case sensitivity.
4. Add a testcase for superuser(if it is not possible to add it in automated tests, please test it manually and add to description.)

**Review by HarshLunagariya — CHANGES_REQUESTED** [review]

File: test/JDBC/expected/restricted_objects-vu-verify.out (hunk 10 above)
Comment (roshan0708) [reply test/JDBC/expected/restricted_objects-vu-verify.out:48]: Tests for other objects which are not included in this PR have already been added to `restricted_procedures` test file. This change is only for `msdb` objects.

File: contrib/babelfishpg_tsql/src/pl_funcs-2.h [hunk 11]
```diff
@@ -3,5 +3,6 @@
 
 static void free_stmt2(PLtsql_stmt *stmt);
 static void dump_stmt2(PLtsql_stmt *stmt);
+extern bool superuser(void);
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/pl_funcs-2.h:6]: Why can't directly include it from miscadmin.h?

**Review by HarshLunagariya — APPROVED** [review]

**Review by shalinilohia50 — APPROVED** [review]

**Review by shardgupta — APPROVED** [review]
